mongoose/mongoose.d.ts

```diff
@@ -269,14 +273,43 @@ declare module "mongoose" {
     path(path: string, constructor: any): Schema;
     pathType(path: string): string;
     plugin(plugin: (schema: Schema, options?: Object) => void, options?: Object): Schema;
-    post(method: string, fn: Function): Schema;
-    pre(method: string, callback: Function): Schema;
+
+    pre(method: string, fn: HookSyncCallback, errorCb?: HookErrorCallback): Schema;
+    pre(method: string, isAsync: boolean, fn: HookAsyncCallback, errorCb?: HookErrorCallback): Schema;
+    post(method: string, fn: HookSyncCallback, errorCb?: HookErrorCallback): Schema;
+    post(method: string, isAsync: boolean, fn: HookAsyncCallback, errorCb?: HookErrorCallback): Schema;
+
     requiredPaths(): string[];
     set(key: string, value: any): void;
     static(name: string, fn: Function): Schema;
     virtual(name: string, options?: Object): VirtualType;
     virtualpath(name: string): VirtualType;
+
+  }
+
+  // hook functions: https://github.com/vkarpov15/hooks-fixed
+  export interface HookSyncCallback {
+      (next: HookNextFunction, ...hookArgs:any[]): any;
+  }
+
+  export interface HookAsyncCallback {
+      (next: HookNextFunction, done: HookDoneFunction, ...hookArgs:any[]): any;
+  }
+
+  export interface HookErrorCallback {
+      (error: Error): any;
+  }
+
+  export interface HookNextFunction {
+      (...hookArgs:any[]): any;
+      (error: Error): any;
```

Comment:
The second argument is not needed. the first one subsumes it. if you want to keep them both, then make the less specific one first.